onnxruntime/core/framework/stream_execution_context.h

```diff
@@ -0,0 +1,171 @@
+class StreamExecutionContext {
+ public:
+  /*
+   * LIMITATION:
+   * CountDownBarrier is only for scenario that the v is set
+   * to the # of consumers and each consumer calls Dec() exactly once.
+   */
+  class CountDownBarrier {
+   public:
+    CountDownBarrier() : v_{0} {};
+
+    void Set(int32_t v) {
+      ORT_ENFORCE(v >= 0);
+      v_.store(v, std::memory_order_relaxed);
+    }
+
+    bool Dec() {
+      return v_.fetch_sub(1, std::memory_order_relaxed) == 1;
+    }
+
+    int32_t Get() { return v_.load(std::memory_order_relaxed); }
+
+    void Inc() {
+      ++v_;
+    }
+
+   private:
+    std::atomic_int_fast32_t v_;
+  };
+
+  StreamExecutionContext(const SessionState& sess_state,
+                         int32_t num_streams,
+                         gsl::span<const size_t> notification_owners,
+                         gsl::span<const int> feed_mlvalue_idxs,
+                         gsl::span<const OrtValue> feeds, gsl::span<const int> fetch_mlvalue_idxs,
+                         std::vector<OrtValue>& fetches,
+                         const InlinedHashMap<size_t, IExecutor::CustomAllocator>& fetch_allocators,
+                         size_t num_barriers,
+                         const logging::Logger& sess_logger,
+                         const DeviceStreamCollection& device_stream_map,
+                         bool single_thread_mode);
+
+  const SessionState& GetSessionState() const;
+
+  const logging::Logger& GetLogger() const;
+
+  ExecutionFrame& GetExecutionFrame();
+
+  synchronize::Notification* GetNotification(size_t idx);
+
+  void SetLogger(const logging::Logger& current_logger) {
+    logger_ = &current_logger;
+  }
+  // Get status of the execution.
+  // if one of the stream got non-OK status, the whole task status will be set as that non-OK status.
+  const Status& TaskStatus() const;
+  // Decrease the count of a given barrier.
+  bool DecCountDownBarrier(size_t barrier_id);
+  // The execution mode:
+  // 1. single thread mode: all the streams will be launched using current host thread.
+  // 2. multi-threads mode: use inter-op thread pool to schedule the N streams.
+  bool SingleThreadMode() const { return single_thread_mode_; }
+  // Get the Stream instance for a given logic sequence.
+  // return nullptr if the device of given logic sequence doesn't register stream support.
+  Stream* GetDeviceStream(size_t idx);
+  // Decrease the count of remaining job by 1.
+  void CompleteTask();
+  // Increase the count of remaining job by 1.
+  void AddTask();
+  // This is only used under multi-threads mode.
+  // blocked until all the jobs scheduled into inter-op thread pool complete.
+  void WaitAll();
+  // If one of the stream got non-OK status, update the status in the context.
+  void SetStatus(Status& status);
+  // Release the OrtValues after a step, based on the execution plan.
+  void RecycleNodeInputs(onnxruntime::NodeIndex node_index);
```

Comment:
nit: can we have a newline between each separate thing so it's easier to read?

Reply:
sure